File: 300.500/476.NumberComplement.cpp
#include <iostream>
using namespace std;

class NumberComplement
{
public:
    int findComplement(int num)
    {
        int highbit = 0;
        for (int i = 1; i <= 30; ++i)
        {
            if (num >= (1 << i))
            {
                highbit = i;
            }
            else
            {
                break;
            }
        }
        // 特殊处理highbit == 30 是防止溢出
        int mask = (highbit == 30 ? 0x7fffffff : (1 << (highbit + 1)) - 1);
        return num ^ mask;
    }
};

int main(int argc, char const *argv[])
{
    NumberComplement cls;
    cout << cls.findComplement(5) << " -> 2" << endl;
    cout << cls.findComplement(1) << " -> 0" << endl;
    return 0;
}

File: go/300.500/476.NumberComplement.go
package main

import (
	"fmt"
)

func findComplement(num int) int {
	h := 0
	for i := 0; i <= 30; i++ {
		if num >= (1 << i) {
			h = i
		} else {
			break
		}
	}
	mask := (1 << (h + 1)) - 1
	return num ^ mask
}

func main() {
	fmt.Println(findComplement(2), 1)
	fmt.Println(findComplement(5), 2)
	fmt.Println(findComplement(1), 0)
}
